build-zones

```diff
@@ -0,0 +1,194 @@
+#!/usr/bin/env python3
+"""Rebuild DNS config based on LDAP and template files.
+
+The eventual goal is to build almost all of DNS from LDAP, including things
+like CNAMEs, and just use this script to generate the zones.
+
+Right now, it only generates A/AAAA and reverse-DNS records for hosts, and the
+result needs to be checked in to git.
+"""
+import re
+from collections import namedtuple
+from dns import reversename
+from ipaddress import ip_address
+from itertools import chain
+from operator import attrgetter
+
+import ldap3
+from cached_property import cached_property
+from ocflib.constants import OCF_LDAP_HOSTS
+from ocflib.infra.net import ipv4_to_ipv6
+from ocflib.infra.net import is_ocf_ip
+from ocflib.infra.ldap import ldap_ocf
+
+
+BANNER = '''\
+; @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
+;                THIS FILE IS AUTOMATICALLY GENERATED
+; @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
+;
+;   You should either make changes in LDAP and run `make`, or modify
+;   the template under `templates`.
+;
+; @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
+'''
+
+
+# TODO: maybe this is a useful thing to put in ocflib?
+class Host(namedtuple('Host', ('hostname',))):
+
+    @classmethod
+    def from_ldap_record(cls, record):
+        attrs = record['attributes']
+        hostname, = attrs['cn']
+        ip, = ip_address(attrs['ipHostNumber'])
+        return cls(
+            hostname=hostname,
+            ip=ip,
+            cnames=set(),
+        )
+
+    @cached_property
+    def description(self):
+        d = '{self.hostname} ({self.ldap_attrs[type][0]})'.format(self=self)
+        if 'puppetClass' in self.ldap_attrs:
+            d += ', roles: [{}]'.format(', '.join(sorted(self.ldap_attrs['puppetClass'])))
+        if self.owner:
+            d += ', owner={}'.format(self.owner)
+        return d
+
+    @cached_property
+    def owner(self):
+        """Return owner, or None."""
+        for pvar in self.ldap_attrs.get('puppetVar', []):
+            match = re.match('^owner=([a-z]{3,16})$', pvar)
+            if match:
+                return match.group(1)
+
+    @cached_property
+    def ipv4(self):
+        attrs = self.ldap_attrs
+        ipv4, = attrs['ipHostNumber']
+        ipv4 = ip_address(ipv4)
+        assert is_ocf_ip(ipv4), ipv4
+        return ipv4
+
+    @cached_property
+    def ipv6(self):
+        ipv6 = ipv4_to_ipv6(self.ipv4)
+        assert is_ocf_ip(ipv6), ipv6
```

Comment:
i don't think it can happen in the current implementation. think of this as a cheap test.